components/ai_chat/core/browser/model_service.cc

```diff
@@ -262,6 +268,63 @@ void ModelService::MigrateProfilePrefs(PrefService* profile_prefs) {
   }
 }
 
+// Custom models do not have fixed properties pertaining to the number of
+// characters they can process before a potential-coherence-loss warning is
+// shown. Leo Models have hard-coded values, but custom models' properties are
+// based on their context size, which may or may not have been provided by the
+// user. For this reason, we set the long_conversation_warning_character_limit
+// and max_associated_content_length after the model has been loaded and
+// validated.
+void ModelService::SetAssociatedContentLengthMetrics(mojom::Model& model) {
+  if (!model.options->is_custom_model_options()) {
+    // Only set metrics for custom models
+    return;
+  }
+
+  if (!ModelValidator::HasValidContextSize(
+          *model.options->get_custom_model_options())) {
+    model.options->get_custom_model_options()->context_size =
+        kDefaultCustomModelContextSize;
+  }
+
+  uint32_t max_associated_content_length =
+      ModelService::GetMaxAssociatedContentLengthForModel(model);
+
+  model.options->get_custom_model_options()->max_associated_content_length =
+      max_associated_content_length;
+
+  base::CheckedNumeric<uint32> warn_at = base::CheckMul<size_t>(
+      max_associated_content_length, kMaxContentLengthThreshold);
+
+  if (warn_at.IsValid()) {
+    model.options->get_custom_model_options()
+        ->long_conversation_warning_character_limit = warn_at.ValueOrDie();
+  }
```

Comment:
why not set max_associated_content_length here too, instead of recalculating it when needed?

Reply:
I set it a few lines up. You're right, we could absolutely avoid the recalc later.